src/Microsoft.ML.Tokenizers/Normalizer/UpperCaseNormalizer.cs

```diff
@@ -22,5 +24,26 @@ public sealed class UpperCaseNormalizer : Normalizer
         /// <param name="original">The original string to normalize to uppercase form.</param>
         /// <returns>The upper-cased normalized string.</returns>
         public override string Normalize(string original) => original.ToUpperInvariant();
+
+        /// <summary>
+        /// Uppercase the original string.
+        /// </summary>
+        /// <param name="original">The original string to normalize to uppercase form.</param>
+        /// <returns>The upper-cased normalized string.</returns>
+        public override string Normalize(ReadOnlySpan<char> original)
+        {
+            if (original.IsEmpty)
+            {
+                return string.Empty;
+            }
+
+            char[] arrayPoolArray = ArrayPool<char>.Shared.Rent(original.Length);
+
+            int length = original.ToUpperInvariant(arrayPoolArray);
+            Debug.Assert(length == original.Length);
+
+            ArrayPool<char>.Shared.Return(arrayPoolArray);
+            return new string(arrayPoolArray, 0, length);
```

Comment:
Is it OK to use this string after returning it to the pool?  Couldn't someone else get it in between the return and the new string copying its value?

Reply:
Ops. this was not intentional. Thanks for the catch. We have the same issue in Lowercase one too :-(